com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java

```diff
@@ -102,14 +112,70 @@ protected void sendMessage(Messages.ProtocolMessage message) {
         String utf8Data = data.getString(PROTOCOL_ENCODING);
 
         try {
-            logger.fine("\n[[RESPONSE]]\n" + utf8Data);
+            if (message instanceof Messages.Request) {
+                logger.fine("\n[[REQUEST]]\n" + utf8Data);
+            } else if (message instanceof Messages.Event) {
+                logger.fine("\n[[EVENT]]\n" + utf8Data);
+            } else {
+                logger.fine("\n[[RESPONSE]]\n" + utf8Data);
+            }
             this.writer.write(utf8Data);
             this.writer.flush();
         } catch (IOException e) {
             logger.log(Level.SEVERE, String.format("Write data to io exception: %s", e.toString()), e);
         }
     }
 
+    /**
+     * Send a request to the DA.
+     *
+     * @param request
+     *            the request message.
+     * @param cb
+     *            the request call back function.
+     */
+    protected void sendRequest(Messages.Request request, Consumer<Messages.Response> cb) {
+        sendRequest(request, 0, cb);
+    }
+
+    /**
+     * Send a request to the DA. And create a timeout error response to the callback if no response is received at the give time.
+     *
+     * @param request
+     *            the request message.
+     * @param timeout
+     *            the maximum time (in millis) to wait.
+     * @param cb
+     *            the request call back function.
+     */
+    protected void sendRequest(Messages.Request request, int timeout, Consumer<Messages.Response> cb) {
+        sendMessage(request);
+        if (cb != null) {
+            pendingRequests.put(request.seq, cb);
+            if (timeout > 0) {
+                CompletableFuture.runAsync(() -> {
+                    try {
+                        Thread.sleep(timeout);
+                        Consumer<Messages.Response> callback = fetchPendingRequestCallback(request.seq);
+                        if (callback != null) {
+                            callback.accept(new Messages.Response(request.seq, request.command, false, "timeout"));
```

Comment:
Why do we construct the response object ourselves? I think vs code is responsible for sending back the response.

Reply:
AbstractProtocolServer is just neutral server code, we need consider the case VSCode will send a normal response back, we also need consider the case that VSCode doesn't give a response back. So the server provides a timeout option.